File: python/2023/02_cube_conundrum/aoc202302.py
"""AoC 2, 2023: Cube Conundrum."""

# Standard library imports
import math
import pathlib
import sys

# Third party imports
import parse

GAME_PATTERN = parse.compile("Game {game:d}: {cubes}")
CUBE_PATTERNS = (
    parse.compile("{num:d} red"),
    parse.compile("{num:d} green"),
    parse.compile("{num:d} blue"),
)


def parse_data(puzzle_input):
    """Parse input."""
    return {
        match["game"]: parse_cubes(match["cubes"])
        for line in puzzle_input.split("\n")
        if (match := GAME_PATTERN.parse(line))
    }


def parse_cubes(cube_descriptions):
    """Parse cubes into tuples, in a (red, green, blue) order

    ## Examples:

    >>> parse_cubes("3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green")
    [(4, 0, 3), (1, 2, 6), (0, 2, 0)]
    """
    return [
        tuple(
            match["num"] if (match := pattern.search(description)) else 0
            for pattern in CUBE_PATTERNS
        )
        for description in cube_descriptions.split(";")
    ]


def part1(data):
    """Solve part 1."""
    bag = (12, 13, 14)
    return sum(
        id
        for id, reveals in data.items()
        if all(all(r <= b for r, b in zip(reveal, bag)) for reveal in reveals)
    )


def part2(data):
    """Solve part 2."""
    return sum(math.prod(minimum_cubes(reveals)) for reveals in data.values())


def minimum_cubes(reveals):
    """Find the minimum number of cubes necessary to support the reveals.

    ## Example:

    >>> minimum_cubes([(0, 4, 4), (1, 5, 2), (1, 0, 9), (0, 0, 1)])
    (1, 5, 9)
    """
    return tuple(max(cubes) for cubes in zip(*reveals))


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2023/02_cube_conundrum/aoc202302.ex
defmodule AOC2023.Day02 do
  @moduledoc """
  Advent of Code 2023, day 2: Cube Conundrum.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.map(&find_max_cubes/1)
  end

  @doc """
  Find the maximum number of cubes in each color.

  Return a tuple in the order {red, green, blue}. We don't need to care about
  the difference between , and ; to find the maximum.

  ## Example:

      iex> find_max_cubes("Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green")
      {6, 3, 2}
  """
  def find_max_cubes(line) do
    [_, cubes] = String.split(line, ":")

    cubes
    |> String.split(~r/[,;]/)
    |> Enum.map(fn cube ->
      [num, color] = String.split(cube)
      {color, String.to_integer(num)}
    end)
    |> Enum.reduce(%{}, fn {color, num}, acc ->
      Map.update(acc, color, num, &max(&1, num))
    end)
    |> then(fn %{"red" => red, "green" => green, "blue" => blue} -> {red, green, blue} end)
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    data
    |> Enum.with_index(1)
    |> Enum.filter(fn {{red, green, blue}, _} -> red <= 12 && green <= 13 && blue <= 14 end)
    |> Enum.map(fn {_, game_id} -> game_id end)
    |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    data
    |> Enum.map(fn {red, green, blue} -> red * green * blue end)
    |> Enum.sum()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
